Fix #74020: Optimize consecutive shifts in JIT Morph

csa7mdm opened the pull request:
## Description
This PR addresses issue #74020 by adding a peephole optimization in the JIT **Morph** phase (`fgMorphSmpOpOptional`) to combine consecutive right shift operations (GT_RSH/GT_RSZ) with constant shift amounts.

Per reviewer feedback from @tannergooding and @JulieLeeMSFT, the optimization was moved from Lowering to Morph so downstream passes (VN, CSE, LSRA) benefit from the simplified IR.

## Changes
**`src/coreclr/jit/morph.cpp`** - Added `case GT_RSH` / `case GT_RSZ` to `fgMorphSmpOpOptional`:
- Fold `(x shift c1) shift c2` into `x shift (c1+c2)` for same-operation shifts
- Handle RSH overshift with saturation to `bitWidth - 1`  
- Handle RSZ overshift by folding to zero constant
- Guarded by `fgGlobalMorph`  

## Verification
- Addresses the scenario in #74020 where repeated `x / c1 / c2` generates redundant `shr` instructions
- Combined shift count validated against type bit width
- Only fires for same-operation shifts (mixed RSH/RSZ excluded for safety)

tannergooding commented on src/coreclr/jit/lower.cpp:
```diff
@@ -8581,6 +8581,65 @@ void Lowering::LowerShift(GenTreeOp* shift)
         shift->gtOp2->ClearContained();
     }
 
+    if (comp->opts.OptimizationEnabled() && shift->OperIs(GT_RSH, GT_RSZ) && shift->gtGetOp2()->IsCnsIntOrI())
```
Why doesn't this handle left-shift?

`(x << 1) << 2` can be `x << 3`

tannergooding reviewed (approved):
LGTM. I do still think this would be better to handle in morph and that it would be good to track `RSH(x, n)` for `0 < n < bitWidth` as being "never negative", but will defer to other input.

CC. @dotnet/jit-contrib for secondary review

csa7mdm committed:
Fix #74020: Optimize consecutive shifts in JIT Morph

Add peephole optimization in fgMorphSmpOpOptional to fold consecutive
same-operation right shifts with constant amounts:
  (x shift c1) shift c2 -> x shift (c1 + c2)

Handles overshift: RSH saturates to bitWidth-1, RSZ folds to zero.
Guarded by fgGlobalMorph, side-effect checks, and multi-reg checks.

Includes comprehensive tests in ShiftCombining.cs covering RSH/RSZ
for int32/int64, overshift saturation, boundary cases, and the
original division pattern from issue #74020.

csa7mdm committed:
Merge branch 'main' into fix/74020-jit-shift-opt

EgorBo commented on src/coreclr/jit/morph.cpp:
```diff
@@ -11301,6 +11301,68 @@ GenTree* Compiler::fgMorphSmpOpOptional(GenTreeOp* tree, bool* optAssertionPropD
+                && !op1->IsMultiRegNode()
+                && !(op1->gtFlags & (GTF_ALL_EFFECT | GTF_SET_FLAGS)))
+            {
+                ssize_t  c1       = op1->AsOp()->gtOp2->AsIntCon()->IconValue();
```
please use gtGetOp2() instead of AsOp()->gtOp2. same for op1. We only use fields when we need to assign new trees to them.

EgorBo commented on src/coreclr/jit/morph.cpp:
```diff
@@ -11301,6 +11301,68 @@ GenTree* Compiler::fgMorphSmpOpOptional(GenTreeOp* tree, bool* optAssertionPropD
+                    if (combined < (ssize_t)bitWidth)
+                    {
+                        // Safe to combine: x shift (c1 + c2)
+                        JITDUMP("Folding consecutive shifts: (x %s %d) %s %d -> x %s %d\n",
```
please remove OpNames from this and other JITDUMP, I don't think this simple transform worth extra logging

EgorBo commented on src/coreclr/jit/morph.cpp:
```diff
@@ -11301,6 +11301,68 @@ GenTree* Compiler::fgMorphSmpOpOptional(GenTreeOp* tree, bool* optAssertionPropD
+                        return tree;
+                    }
+                    else
+                    {
```
let's remove this path

EgorBo commented on src/tests/JIT/opt/InstructionCombining/ShiftCombining.cs:
```diff
@@ -0,0 +1,238 @@
+    {
+        // (x >> 3) >> 4  ->  x >> 7
+        bool pass = true;
+        pass &= Check((Shr32(100, 3) >> 4), 100 >> 7, "RSH combine 3+4 int32");
```
I don't see how any of these tests test the actual optimization. these methods have NoInlining on them. but if you remove it, then it will be constant-folded long before your opt

csa7mdm commented:
Thanks @EgorBo for the review!

Addressed both concerns:

1. **Side-effects**: Added `!(op1->gtFlags & (GTF_ALL_EFFECT | GTF_SET_FLAGS))` and `!op1->IsMultiRegNode()` guards. The folding now bails out if the inner shift subtree has any side effects.

2. **Test coverage**: Added `ShiftCombining.cs` in `src/tests/JIT/opt/InstructionCombining/` with tests covering RSH/RSZ combining (int32/int64), overshift saturation/zero, boundary cases, and the original `x / 4 / 8` division pattern from #74020.

**Justification**: The pattern appears when C# code does repeated integer division by powers of two (`x / c1 / c2`), which lowers to consecutive right shifts.

csa7mdm committed:
Merge branch 'main' into fix/74020-jit-shift-opt
